Chore: Add a proposal document (#3)
m-mohr/geoparquet-viewer: proposal.md

---

## 2. Feature Requirements

| #  | Feature                                                                                     | Priority     |
|----|---------------------------------------------------------------------------------------------|--------------|
| 1  | Load files from local filesystem or publicly hosted URLs                                    | Must         |
| 2  | Map visualization of any geometry type (Point, Line, Polygon, Multi*)                       | Must         |
| 3  | Table visualization of non-geometry columns                                                 | Must         |
| 4  | Full Parquet metadata display: KVP metadata, schema, column types, compression              | Must         |
| 5  | GeoParquet 1.0 support; 1.1 and 2.0 as bonus                                                | Must / Bonus |
| 6  | Bidirectional selection: table row ↔ map geometry                                           | Must         |

> | 7  | Row-group based loading — initial page of 1000 rows, Load More / Load All on demand         | Must         |

Reviewer: ```suggestion
| 7  | Paginated display— initial page of e.g. 1000 rows, Load More / Load All on demand         | Must         |
```

|------------------|----------------------------------------------------|--------------------------------------------|
| Parquet parsing  | `hyparquet` + `hyparquet-compressors` + `hysnappy` | JS-only, runs on main thread               |
| Map renderer     | OpenLayers (`ol`)                                  | SVG/Canvas, CPU-based rendering            |
| Base map tiles   | OL OSM TileLayer                                   |                                            |
| Geometry parsing | OL `WKB` format reader                             | WKB → OL Feature objects                   |
| Reprojection     | `proj4js` + OL proj4 register                      | Fetched WKT defs from spatialreference.org |
| State            | Vue Options API `data()` in `App.vue`              | Monolithic — all logic in one file         |
| Pagination       | hyparquet `onChunk` row-group streaming            | 100 rows per page                          |

### What worked well

- hyparquet's `asyncBuffer` pattern (HTTP Range requests) was elegant for streaming
- OpenLayers' native any-CRS tile reprojection worked correctly out of the box

### Pain points that drove the migration

> - **All logic in** `App.vue` — 447 lines mixing data loading, map management, WKB parsing, projection, selection, and UI state with no separation of concerns

Reviewer: ```suggestion
```

Irrelevant here.

- **Main-thread blocking** — hyparquet runs synchronously on the main thread; large files freeze the UI
- **No SQL layer** — filtering required manual JS array operations; no aggregation capability
- **OL rendering performance** — OpenLayers' Canvas/SVG renderer struggles beyond ~50k features; large polygon datasets are slow
- **No GeoArrow support** — hyparquet returns plain JS arrays, not columnar Arrow data; no path to GPU-accelerated columnar rendering

---

> ## 4. Explored Options & Why They Were Scrapped

Reviewer: ```suggestion
## 4. Explored Options & Why They Were Discarded
```

**Why it was scrapped:**

- **Reprojection destroys the native CRS experience.** Feature requirement #8 explicitly asks to render data in its native projection. Always converting everything to EPSG:4326 means the basemap is always Web Mercator and the data loses its native coordinate frame. For datasets authored in national grid systems (e.g. ETRS89 / EPSG:25832 in Germany, OSGB36 in the UK), this is a meaningful loss of fidelity for GIS users.
- **MapLibre cannot reproject its own basemap tiles** to arbitrary CRS. Even if deck.gl rendered the geometry correctly in a non-Mercator space, the MapLibre tile layer underneath would be geometrically wrong. This was a hard blocker.
- **The GeoArrow pipeline is preserved** in the final architecture, so the performance advantage of this option is not lost — only the MapLibre basemap is replaced.

> ---

Reviewer: No mention of OpenLayers? Generally, this section is all over the place. Somtimes it's about just a mapping library, sometimes about the whole flow from reading to displaying.

Author: Unless if I am missing something, but this option is talking about what you have on the new branch(AI generated) you worked on and that doesn't use openLayers.

Reviewer: The title of the chapter is "Explored Options & Why They Were Scrapped" and we did explore OpenLayer as well...

## 5. Chosen Architecture

**DuckDB-Wasm (data + SQL) + OpenLayers (basemap, any-CRS tile reprojection) + deck.gl (GPU geometry rendering as OL overlay)**

This hybrid combines:

- **DuckDB-Wasm** for all Parquet reading, metadata extraction, SQL filtering, pagination, and geometry encoding detection.

> - **OpenLayers** for basemap tile management and native any-CRS tile reprojection. OL has first-class proj4js integration and can reproject OSM/WMTS tiles to any registered EPSG code client-side. This is exactly what it was built for.

Reviewer: Why raster tiles?

Author: Update to using Vector tiles.
